streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2424,6 +2452,61 @@ object ZStream {
       }
   }
 
+  private[stream] sealed abstract class Structure[-R, +E, +A] {
+    def process: ZManaged[R, E, ZIO[R, Option[E], A]]
+  }
+
+  private[stream] object Structure {
+    final case class Iterator[-R, +E, +A](process: ZManaged[R, E, Pull[R, E, A]]) extends Structure[R, E, A]
+    final case class Concat[-R, +E, +A](hd: Structure[R, E, A], tl: () => Structure[R, E, A])
+        extends Structure[R, E, A] {
+      val process: ZManaged[R, E, Pull[R, E, A]] = {
+        def go(
+          doneRef: Ref[Boolean],
+          currPull: Ref[Pull[R, E, A]],
+          nextPull: Ref[Option[() => Structure[R, E, A]]],
+          switchPull: ZManaged[R, E, Pull[R, E, A]] => ZIO[R, E, Pull[R, E, A]]
+        ): ZIO[R, Option[E], A] =
+          doneRef.get.flatMap { done =>
+            if (done) IO.fail(None)
+            else
+              currPull.get.flatten.foldM(
+                {
+                  case e @ Some(_) => ZIO.fail(e)
+                  case None =>
+                    nextPull.get.flatMap {
+                      case None => doneRef.set(true) *> IO.fail(None)
+                      case Some(tl) =>
+                        tl() match {
+                          case Iterator(iter) =>
+                            switchPull(iter).mapError(Some(_)).tap(currPull.set) *>
+                              nextPull.set(None) *> go(doneRef, currPull, nextPull, switchPull)
+                          case Concat(hd, tl) =>
+                            // It is extremely important in this case to *NOT* recurse using
+                            // tl.process, as that would introduce a new ZManaged scope; this
+                            // will cause space leaks when used with streams that concatenate infinitely.
+                            // Instead, we re-use the same ZManaged scope introduced by the ZManaged.switchable
+                            // below to swap the current stream being pulled with tl.
+                            switchPull(hd.process).mapError(Some(_)).tap(currPull.set) *>
+                              nextPull.set(Some(tl)) *> go(doneRef, currPull, nextPull, switchPull)
+                        }
```

Comment:
I'm actually going to avoid pulling more things up into `Structure` because I think this is going to end up as a mini-algebra, with an additional `FlatMap` constructor.

Then I'll move the interpreter up from `Concat` (which is the only composite term requiring interpretation) into `Structure`.